rfcs/0024-hard-forks.md

```diff
@@ -0,0 +1,455 @@
+parameters that are set to the same value in all currently supported eras (now +
+the next one) and change the types of the associated fields to directly include
+the values.
+
+#### Handling changes in types
+
+Different kinds of changes will have to be implemented differently. If types
+from a new era only add optional fields, or new fields that have a reasonable
+default, then we can convert incoming data from era n into the era n+1 version,
+and only store the items with the type parameterization of the current era. If
+this isn't possible then we will have to either have both options in an
+`Either.t` or in some cases two parallel data structures.
+
+##### Blocks + block components
+
+If there is no or a small enough change to the consensus mechanism then the
+transition frontier can have blocks of both types, or blocks of only the new
+type with conversion. A sufficiently different consensus mechanism might need a
+separate transition frontier or other data structure to support it. In that
+case, blocks from era n will go in the old transition frontier and blocks from
+era n+1 will go into the new structure. The era transition block will be a
+special case.
+
+##### User transactions
+
+As of now our rule is that we only accept transactions into the pool if they
+could be applied to the best tip. Continuing in this vein, we will only accept
+transactions into the pool if they are valid in the current era. This will cause
+a small hiccup in throughput when era transitions happen, but that's acceptable
+since they are rare.
+
+##### Ledger SNARKs
+
+We also do not accept ledger SNARKs until we've entered the appropriate era for
+them and there is a block whose children might need to buy them.
+
+#### Signaling support for the next era
+
+We would like to have a mechanism that lets users know if they're out of date,
+and forces them to update/informs them they are or will soon be on a dead
+network. Block producers could include the next era id and transition slot in
+every block, but there is no incentive to be truthful, so I'm not convinced the
+feature is actually all that useful. Checking an HTTP endpoint operated by O(1)
+would work, though it's more centralized than I'd like. (Coda needs to continue
+working if everyone who works here dies and the Bay Area falls into the sea.)
+
+We could also have accounts or block producers vote on-chain whether to move to
+a new era. End users will generally prefer the system to be hands-off, and it
+simplifies development not to have it.
+
+#### Parallel scan transitions
+
+It would be easiest to deal with changes to the parallel scan state or the
+SNARKs within it if it were drained and restarted cleanly when necessary. This
+would simplify things a lot and mean that the merge SNARKs didn't have to be
+able to verify both old and new proofs. Below is a scheme to do this. The
+disadvantage is that it requires effectively pausing progress while this
```

Comment:
I think this may not be an acceptable cost. Could we spec some alternative for working with scan states without draining?

Reply:
To be clear, it’s because throughput will drop to a crawl for the time it takes to drain the scan state which is likely to not be short since we don’t do dynamic resizing